### PR TITLE
Add resource naming breaking change for .NET Core 3.0

### LINKED ISSUE #16964: MSBuild Resource Naming Convention Breaking Change
## Resource Naming Conventions Have Changed
For those that would like a slightly higher level explanation and maybe even a mediocre diagram of what's going on here, see [this gist](https://gist.github.com/BenVillalobos/041673b9a73bec60fdc3bf0f86fae62a) I created.

### Version introduced
3.0, via [applying DependentUpon convention](https://github.com/microsoft/msbuild/pull/4597), but more specifically [applying that convention to localized resources](https://github.com/microsoft/msbuild/pull/4853). Here is the [related MSBuild issue](https://github.com/microsoft/msbuild/issues/4996) that tracked this.

### Old behavior
When the `DependentUpon` metadata was set, the manifest name would be `FullTypeName.resources` where FullTypeName is `Namespace.Classname`. When the `DependentUpon` metadata was NOT set, the manifest names default to: `FullTypeName.FolderPathRelativeToRoot.Culture.resources`. 
    

### New behavior
When a resx is associated with a type like in the winforms case, the generally presumed convention is that the resource is called `FullTypeName.resources` (the culture may or may not be present, `FullTypeName.Culture.resources` for example).

From this presumed convention, the rules around how RootNamespace and folders impact the name chosen mirror how VS picks a namespace and name for a new class: `RootNamespace.FoldersSeparatedByDots.FileName.(resx|resources|cs)`

So if you leave all the defaults when you add files and resources in VS, then when you have Type.cs next to Type.resx, you will get FullNameOfType.resources. However, when you stray from the defaults, there’s more to it.

#### 1. If the metadata LogicalName `%(EmbeddedResource.LogicalName)` is set, then use it as is.
Examples:
```xml
      <EmbeddedResource Include="X.resx" LogicalName="SomeName.resources" />
      <EmbeddedResource Include="X.fr-FR.resx" LogicalName="SomeName.resources" />
```
Then the resource is `SomeName.resources`, irrespective of the `.resx` file name, culture, or any other metadata.

#### 2. else if the metadata ManifestResourceName `%(EmbeddedResource.ManifestResourceName)` is set then add the `.resources` extension and use it.
   
Examples: 
```xml
<EmbeddedResource Include="X.resx" ManifestResourceName="SomeName" />
```
Then the resource name is SomeName.resources
```xml
<EmbeddedResource Include="X.fr-FR.resx" ManifestResourceName="SomeName.fr-FR" />
```
Then the resource name is SomeName.fr-FR.resources

#### 3. else if `%(EmbeddedResource.DependentUpon)` is set to a source file, then parse the source file and add culture and `.resources` extension to the full type name of the first class in the file. This allows types to be renamed and have the corresponding resource change names accordingly.

Given Q.cs:

      namespace SomeNamespace { class SomeType {} }}
Examples:
```xml
<EmbeddedResource Include="X.resx" DependentUpon="Q.cs">
```
Then the resource name is `SomeNamespace.SomeType.resources`
```xml
<EmbeddedResource Include="X.fr-FR.resx" DependentUpon="Q.cs">
```
Then the resource name is `SomeNamespace.SomeType.fr-FR.resources`

#### 4. else if `EmbeddedResourceUseDependentUponConvention` is set to `true` (**which is true by default for projects targeting .net core 3+**), then choose DependentUpon as follows and perform rule 3 above.

If there is a source file in the same folder that has the same name as the resx when not including culture or file extensions, choose DependentUpon from the resx to the source file.

Example: 

&nbsp;&nbsp;Src/
&nbsp;&nbsp;&nbsp;&nbsp;Q.cs: namespace SomeNamespace { class SomeType {} }}
&nbsp;&nbsp;&nbsp;&nbsp;Q.resx --> resource name is `SomeNamespace.SomeType.resources`
&nbsp;&nbsp;&nbsp;&nbsp;Q.fr-Fr.resx --> resource name is `SomeNamespace.SomeType.fr-FR.resources`

#### Reason for change
We had various issues([1](https://github.com/dotnet/winforms/issues/638), [2](https://github.com/dotnet/project-system/issues/4807), [3](https://github.com/dotnet/corefx/issues/34190)) revolving around the desire to get users to migrate to .NET Core and how MSBuild handles embedded resources being a blocker for that.

#### Recommended action
If you're unsatisfied with the manifest names that are currently being generated, simply add this line to your project file:

    <EmbeddedResourceUseDependentUponConvention>false</EmbeddedResourceUseDependentUponConvention>

Setting this tag to false is how you opt out of this convention entirely. Do note that rules 1 and 2 above **still apply**.

Otherwise, modify your resource file metadata to satisfy one of the above rules.

#### Category

- ASP.NET Core
- C#
- MSBuild
- Windows Forms
- Windows Presentation Foundation (WPF)

---
#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #16964

[Internal preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/msbuild?branch=pr-en-us-17044#resource-manifest-file-names).

### CODE REVIEW

File: includes/core-changes/msbuild/3.0/resource-manifest-name.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+Starting in .NET Core 3.0, the generated resource manifest file name has changed.
+
+#### Version introduced
+
+3.0
+
+#### Change description
+
+Prior to .NET Core 3.0, when [DependentUpon](/visualstudio/msbuild/common-msbuild-project-items#compile) metadata was set for a resource file in the MSBuild project file, the generated manifest name was *Namespace.Classname.resources*. When [DependentUpon](/visualstudio/msbuild/common-msbuild-project-items#compile) was not set, the generated manifest name was Namespace.Classname.FolderPathRelativeToRoot.Culture.resources.
```
Comment (tdykstra) [inline includes/core-changes/msbuild/3.0/resource-manifest-name.md]:
```suggestion
Prior to .NET Core 3.0, when [DependentUpon](/visualstudio/msbuild/common-msbuild-project-items#compile) metadata was set for a resource file in the MSBuild project file, the generated manifest name was *Namespace.Classname.resources*. When [DependentUpon](/visualstudio/msbuild/common-msbuild-project-items#compile) was not set, the generated manifest name was *Namespace.Classname.FolderPathRelativeToRoot.Culture.resources*.
```

File: includes/core-changes/msbuild/3.0/resource-manifest-name.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+Starting in .NET Core 3.0, when a *.resx* file is associated with a type, for example, in Windows Forms apps, it's assumed that the resource is named *Namespace.Classname\[.Culture].resources*. From this convention, the rules around how the root namespace and folders impact the chosen name mirror how Visual Studio picks a namespace and name for a new class: *RootNamespace.FoldersSeparatedByDots.FileName.(resx|resources|cs)*.
```
Comment (tdykstra) [inline includes/core-changes/msbuild/3.0/resource-manifest-name.md]: The sentence beginning "From this convention..." is hard to follow.

File: includes/core-changes/msbuild/3.0/resource-manifest-name.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+If you leave all the defaults when you add files and resources in Visual Studio, then when *Type.cs* is colocated with *Type.resx*, the generated manifest name is *Namespace.Classname.resources*.
+
+If you don't use the default settings, there are more considerations as to how the manifest name is generated:
+
+- If the `LogicalName` attribute on the `EmbeddedResource` element is set, that value is used as the resource file name.
+
+  Examples:
+
+  `<EmbeddedResource Include="X.resx" LogicalName="SomeName.resources" />`
```
Comment (tdykstra) [inline includes/core-changes/msbuild/3.0/resource-manifest-name.md]: These two lines would look better in a fenced code block, or with a blank line between them.

File: includes/core-changes/msbuild/3.0/resource-manifest-name.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  `<EmbeddedResource Include="X.fr-FR.resx" LogicalName="SomeName.resources" />`
+
+  Generated resource name: *SomeName.resources* (regardless of the *.resx* file name or culture or any other metadata)
```
Comment (tdykstra) [inline includes/core-changes/msbuild/3.0/resource-manifest-name.md]: The style guide prescribes periods at end of bulleted text unless it's very short text, i.e., 2-3 words.

File: includes/core-changes/msbuild/3.0/resource-manifest-name.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    Generated resource name: *Namespace.Classname.resources*,(where Namespace.Classname is the name of the first class in *MyTypes.cs*)
+
+  - `<EmbeddedResource Include="X.fr-FR.resx" DependentUpon="MyTypes.cs">`
+
+    Generated resource name: *Namespace.Classname.fr-FR.resources* (where Namespace.Classname is the name of the first class in *MyTypes.cs*)
+
+- If none of the previous rules apply, and `EmbeddedResourceUseDependentUponConvention` is `true` (which is the default value for projects that target .NET Core 3.0 and later), and there's a source file and a *.resx* file with the same base file name in the same folder, the *.resx* file is made dependent upon the matching source file, and the generated name is the same as in the previous rule.
+
+  Example:
```
Comment (tdykstra) [inline includes/core-changes/msbuild/3.0/resource-manifest-name.md]:
```suggestion
  Examples:
```

**Review by tdykstra — APPROVED** [review]
LGTM

**Comment by gewarren** [comment]
@BenVillalobos Could you take a look and see if this info is accurate?

https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/msbuild?branch=pr-en-us-17044

**Comment by BenVillalobos** [comment]
It's a late review but the only changes I had would have been small nits. Looks good!

**Comment by gewarren** [comment]
> It's a late review but the only changes I had would have been small nits. Looks good!

Thanks for reviewing! Feel free to submit a PR for the nits if you like.
